aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,127 @@
 #include <memory>
 #include <stddef.h>
 
-#include "ATen/Retainable.h"
+#include <ATen/Error.h>
+#include <ATen/Retainable.h>
 
 namespace at {
 
+// Note [Separated Allocator and Deleter]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Why are Allocator and Deleter put in separate classes?  The key is that
+// an allocator may allocate a distinct context for every deleter.  This is
+// especially important upon reallocation: if we do not allocate a new
+// context, the contexts of the new and old data can clobber each other.
+// Imagine the following sequence of events:
+//
+//  1. Storage has some data and a BoundAllocatorDeleter associated with it.
+//     The context in this case is an owning reference to "IOBuf", an
+//     enclosing struct for the data.
+//
+//  2. A resize occurs on storage.  We call BoundAllocatorDeleter to
+//     allocate some new memory to store the resized data.  To allocate
+//     this new memory, we must allocate a new IOBuf.  But how can
+//     we update the context to replace the old reference with
+//     the new one?  Disaster!
+//
+// Previously, this case was worked around by directly supporting realloc()
+// in the deleter.  But this is bad for different reasons (it assumes the
+// allocator knows how to copy data; not a safe assumption, since allocators
+// don't know what data is actually contained within them.)
+
+struct Deleter {
+  virtual ~Deleter() {}
+  virtual void deallocate(void* ctx, void* ptr) const = 0;
+};
+
+// WARNING: BoundDeleter may LEAK ctx_ if you never actually call it on
+// the pointer it's supposed to delete; e.g., ctx_'s lifetime may be the
```

Comment:
I feel like I'm missing something here -- is the thing following the e.g. supposed to be an example, or did you mean i.e.?

Reply:
Here's the rewrite:

```
// WARNING: A common pattern when writing BoundDeleter is to set things
// up so that the lifetime of ctx_ is the same as pointer.  In this case,
// you will LEAK ctx_ if you never actually call it on the pointer it's
// supposed to delete.  Use the deleter that comes with a pointer to
// deallocate it; don't do it some out-of-band way!
```